probe/tag/tagger.go

```diff
@@ -0,0 +1,25 @@
+package tag
+
+import "github.com/weaveworks/scope/report"
+
+// Tagger tags nodes with value-add node metadata.
+type Tagger interface {
+	Tag(r report.Report, ts report.TopologySelector, id string) report.NodeMetadata
```

Comment:
This interface is foundational to the new data model. A tagger takes a specific node ID, and some surrounding context to allow it to cross-reference keys and dig up more information, and produces arbitrary metadata that can be merged into that node. This is the primary mechanism to enrich nodes...

> the docker_tagger is going to become much more than just a tagger (its going to have to emit at least one topology itself)

We're talking about two distinct things here:

- The docker tagger is the thing that takes already existing nodes that have process IDs in their metadata, and enriches them with additional metadata taken from Docker
- The container topology producer will be the thing that produces the container topology with information from the Docker daemon

They both may be built on the same Docker event subscriber, but they're separate logical components, nothing really to do with each other in a component model of the probe.

> Finally, the interface (report, selector, id) seems weird. It seems a bit tortuous, and is only ever called in a loop by tagger.

It's a bit awkward only because we need a way to cross-reference metadata across different topologies. For that, we need access to both the node we're enriching, i.e.

```go
md := ts(t).NodeMetadata[id]
```

and a way for the thing that implements the interface to access other, arbitrary metadata from the report, e.g.

```
foreignKey, ok := md["some_key"]
if !ok {
    return NodeMetadata{} // nothing found
}
value := r.Container.NodeMetadatas[foreignKey]["other_key"]
return NodeMetadata{"some_container_data": value}
```

> I suggest for topology tagger, we move the loop into the tagger. For docker tagger, we make it take the process topology explicitly. And we get rid of this interface.

Arguably the whole point of #123 is to generalize the concept of a tagger to something that implements this interface (or one like it). So — if, after this explanation, you still want to do this, I guess we're missing some fundamental conceptual thing... let me know...